File: stock_platform/java/backend/java/src/main/java/com/stockplatform/service/StockPriceService.java
package com.stockplatform.service;

import com.my.mybatis.flex.core.condition.QueryCondition;
import com.stockplatform.dto.StockPriceHistoryDTO;
import com.stockplatform.dto.StockPriceResponseDTO;
import com.stockplatform.entity.Position;
import com.stockplatform.entity.StockPrice;
import com.stockplatform.mapper.PositionMapper;
import com.stockplatform.mapper.StockPriceMapper;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import java.time.LocalDateTime;
import java.util.*;
import java.util.stream.Collectors;

@Service
public class StockPriceService {

    private final StockPriceMapper stockPriceMapper;
    private final PositionMapper positionMapper;
    private final RestTemplate restTemplate = new RestTemplate();

    public StockPriceService(StockPriceMapper stockPriceMapper, PositionMapper positionMapper) {
        this.stockPriceMapper = stockPriceMapper;
        this.positionMapper = positionMapper;
    }

    public StockPriceResponseDTO fetchAndSavePrice(String stockCode) {
        Map<String, Object> data = fetchStockPriceFromApi(stockCode);
        if (data == null || data.get("price") == null) {
            return null;
        }

        StockPrice stockPrice = new StockPrice();
        stockPrice.setStockCode(stockCode);
        stockPrice.setPrice((Double) data.get("price"));
        stockPrice.setChangeValue((Double) data.get("changeValue"));
        stockPrice.setChangePercent((Double) data.get("changePercent"));
        stockPrice.setTradeDate(LocalDateTime.now());
        stockPrice.setCreatedAt(LocalDateTime.now());
        stockPriceMapper.insertSelective(stockPrice);

        return StockPriceResponseDTO.fromEntity(stockPrice);
    }

    public int updateAllPositionsPrice() {
        List<Position> positions = positionMapper.selectAll();
        Set<String> stockCodes = positions.stream()
                .map(Position::getStockCode)
                .collect(Collectors.toSet());

        int count = 0;
        for (String code : stockCodes) {
            if (fetchAndSavePrice(code) != null) {
                count++;
            }
            try {
                Thread.sleep(500);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
        return count;
    }

    @Scheduled(cron = "0 0 16 * * *")
    public void scheduledUpdatePrices() {
        updateAllPositionsPrice();
    }

    public StockPriceResponseDTO getLatestPrice(String stockCode) {
        QueryCondition condition = QueryCondition.create()
                .where("stock_code = ?", stockCode)
                .orderBy("trade_date", true);
        List<StockPrice> prices = stockPriceMapper.selectAllByCondition(condition);
        if (prices.isEmpty()) {
            return null;
        }
        return StockPriceResponseDTO.fromEntity(prices.get(0));
    }

    public StockPriceHistoryDTO getPriceHistory(String stockCode, int days) {
        LocalDateTime startDate = LocalDateTime.now().minusDays(days);
        QueryCondition condition = QueryCondition.create()
                .where("stock_code = ?", stockCode)
                .where("trade_date >= ?", startDate)
                .orderBy("trade_date", true);
        List<StockPrice> prices = stockPriceMapper.selectAllByCondition(condition);

        StockPriceHistoryDTO historyDTO = new StockPriceHistoryDTO();
        historyDTO.setStockCode(stockCode);
        historyDTO.setPrices(prices.stream()
                .map(StockPriceResponseDTO::fromEntity)
                .collect(Collectors.toList()));
        return historyDTO;
    }

    private Map<String, Object> fetchStockPriceFromApi(String stockCode) {
        try {
            String emCode = getEastmoneyCode(stockCode);
            String url = "https://push2.eastmoney.com/api/qt/stock/get?secid=" + emCode + "&fields=f43,f44,f57,f58";

            String response = restTemplate.getForObject(url, String.class);

            Map<String, Object> result = new HashMap<>();
            // 简单解析 JSON - 实际应该用 Jackson 或 Gson
            // 这里仅作示例，实际需要更完善的实现
            // 暂时返回 null，后续可以添加完整的 JSON 解析
            return null;
        } catch (Exception e) {
            return null;
        }
    }

    private String getEastmoneyCode(String stockCode) {
        if (stockCode.startsWith("6")) {
            return "1." + stockCode;
        } else if (stockCode.startsWith("0") || stockCode.startsWith("3")) {
            return "0." + stockCode;
        }
        return stockCode;
    }
}

File: stock_platform/python/backend/python/services/stock_price_service.py
import requests
import time
from models import StockPrice, Position
from datetime import datetime, timedelta
from typing import Optional


def get_eastmoney_code(stock_code: str) -> str:
    if stock_code.startswith("6"):
        return f"1.{stock_code}"
    elif stock_code.startswith(("0", "3")):
        return f"0.{stock_code}"
    return stock_code


def fetch_stock_price(stock_code: str, retries: int = 3) -> Optional[dict]:
    for i in range(retries):
        try:
            em_code = get_eastmoney_code(stock_code)
            url = f"https://push2.eastmoney.com/api/qt/stock/get"

            params = {
                "secid": em_code,
                "fields": "f43,f44,f57,f58"
            }

            response = requests.get(url, params=params, timeout=10)
            data = response.json()

            if data.get("data") and data["data"].get("f43"):
                price = float(data["data"]["f43"]) / 100
                prev_close = float(data["data"]["f44"]) / 100 if data["data"].get("f44") else price
                change_value = price - prev_close
                change_percent = change_value / prev_close * 100 if prev_close else 0

                return {
                    "price": price,
                    "change_value": change_value,
                    "change_percent": change_percent,
                    "trade_date": datetime.now()
                }
            else:
                print(f"股票 {stock_code} 数据格式异常: {data}")
                return None

        except Exception as e:
            print(f"获取股票 {stock_code} 价格失败 (尝试 {i+1}/{retries}): {e}")
            if i < retries - 1:
                time.sleep(1)
    return None


async def fetch_and_save_price(stock_code: str) -> Optional[StockPrice]:
    data = fetch_stock_price(stock_code)
    if not data or data["price"] is None:
        return None

    return await StockPrice.create(
        stock_code=stock_code,
        price=data["price"],
        change_value=data.get("change_value"),
        change_percent=data["change_percent"],
        trade_date=data["trade_date"]
    )


async def update_all_positions_price() -> int:
    positions = await Position.all()
    stock_codes = set(pos.stock_code for pos in positions)

    count = 0
    for code in stock_codes:
        result = await fetch_and_save_price(code)
        if result:
            count += 1
        time.sleep(0.5)

    return count


async def get_latest_price(stock_code: str) -> Optional[StockPrice]:
    price = await StockPrice.filter(stock_code=stock_code).order_by("-trade_date").first()
    return price


async def get_price_history(stock_code: str, days: int = 30) -> list[StockPrice]:
    start_date = datetime.now() - timedelta(days=days)
    return await StockPrice.filter(
        stock_code=stock_code,
        trade_date__gte=start_date
    ).order_by("trade_date")